### PR TITLE
Initial cleanups and tuneups for Terminal-Print

### PR DESCRIPTION
**Pull request by japhb**
A dozen small commits all focused on minor cleanups, refactorings, bug fixes, etc.  A few of the commits also reduce startup time by loading more work into the precompilation phase without losing runtime correctness.

### CODE REVIEW

**Comment by ab5tract** [comment]
Thank you, japhb! These are really helpful commits and much appreciated.

**Comment by ab5tract** [comment]
BTW, I'm hoping to have some time soon to look into writing some lower level functions in C that could help make some speed improvements as well.

**Comment by japhb** [comment]
You're quite welcome.  I sent you another pull request with more.  :-)

On 2016-10-16 09:01, ab5tract wrote:

> Thank you, japhb! These are really helpful commits and much appreciated. 
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or mute the thread [2].
 

Links:
------
[1]
https://github.com/ab5tract/Terminal-Print/pull/12#issuecomment-254055427
[2]

**Comment by japhb** [comment]
Actually, before doing that, I think there may be advantage to making an
algorithmic improvement for cases of printing a lot of text (or drawing
rectangles or similar shapes) -- switching from a column-major to a 
row-major representation in the grid, and taking advantage of available
speedups when rendering horizontal spans of text with the same color (no
need to move the cursor to every cell individually and color each cell
individually, just move to the start of the span and set the color once
for the entire span, and then print a line of text trimmed to the screen
edge). 

To really make use of the available performance gains, this would
require more cleverness in the span handling ... but changing the layout
to be row-major can be done *now* and get modest performance
improvements, while leaving the option of adding cleverness later
without another API break.

On 2016-10-16 09:30, ab5tract wrote:

> BTW, I'm hoping to have some time soon to look into writing some lower level functions in C that could help make some speed improvements as well. 
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or mute the thread [2].
 

Links:
------
[1]
https://github.com/ab5tract/Terminal-Print/pull/12#issuecomment-254057271
[2]

**Comment by ab5tract** [comment]
@japhb , in case you haven't seen yet, I provided a response over at #15 .

tl;dr -- Sounds good to me, as long as the higher level syntax is still row-major. Curious to hear your thoughts!
